x/ibctransfer/onsend_ack.go

```diff
@@ -0,0 +1,95 @@
+	transfertypes "github.com/cosmos/cosmos-sdk/x/ibc/applications/transfer/types"
+	channeltypes "github.com/cosmos/cosmos-sdk/x/ibc/core/04-channel/types"
+
+	sctransfertypes "github.com/Sifchain/sifnode/x/ibctransfer/types"
+	tokenregistrytypes "github.com/Sifchain/sifnode/x/tokenregistry/types"
+)
+
+// OnAcknowledgementMaybeConvert() runs on acknowledgement from receiving chain, of an outgoing send.
+// I.e: Receiving ack from recipient chain of our send.
+// Case A: Recipient acknowledges our send, do nothing.
+// Case B: Recipient acknowledges error of our send, refund sender their coins.
+func OnAcknowledgementMaybeConvert(
+	ctx sdk.Context,
+	sdkTransferKeeper sctransfertypes.SDKTransferKeeper,
+	whitelistKeeper tokenregistrytypes.Keeper,
+	bankKeeper transfertypes.BankKeeper,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+) (*sdk.Result, error) {
+
+	var ack channeltypes.Acknowledgement
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet acknowledgement: %v", err)
+	}
+
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	// OnAcknowledgementPacket responds to the the success or failure of a packet
+	// acknowledgement written on the receiving chain. If the acknowledgement
+	// was a success then nothing occurs. If the acknowledgement failed, then
+	// the sender is refunded their tokens using the refundPacketToken function.
+	if err := sdkTransferKeeper.OnAcknowledgementPacket(ctx, packet, data, ack); err != nil {
+		return nil, err
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			transfertypes.EventTypePacket,
+			sdk.NewAttribute(sdk.AttributeKeyModule, transfertypes.ModuleName),
+			sdk.NewAttribute(transfertypes.AttributeKeyReceiver, data.Receiver),
+			sdk.NewAttribute(transfertypes.AttributeKeyDenom, data.Denom),
+			sdk.NewAttribute(transfertypes.AttributeKeyAmount, fmt.Sprintf("%d", data.Amount)),
+			sdk.NewAttribute(transfertypes.AttributeKeyAck, fmt.Sprintf("%v", ack)),
+		),
+	)
+
+	switch resp := ack.Response.(type) {
+	case *channeltypes.Acknowledgement_Result:
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				transfertypes.EventTypePacket,
+				sdk.NewAttribute(transfertypes.AttributeKeyAckSuccess, string(resp.Result)),
+			),
+		)
+	// if acknowledgement error then a refund was processed so we must check if conversion is necessary
+	case *channeltypes.Acknowledgement_Error:
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				transfertypes.EventTypePacket,
+				sdk.NewAttribute(transfertypes.AttributeKeyAckError, resp.Error),
+			),
+		)
+		// TODO: Always refund, not only if sender of ack is source.
```

Comment:
Should this conversion logic only apply when sender is source? Does the same apply in the other direction? Is this a requirement we need in the code or just something the registry indicates?

Reply:
The sdk's refund logic is triggered above at:
`	if err := sdkTransferKeeper.OnAcknowledgementPacket(ctx, packet, data, ack); err != nil {
		return nil, err
	}`
So a refund is already triggered regardless of whether sender is source. But we need to check for conversion if sender is source because the tokens that were refunded by the sdk function may have been converted tokens. The only tokens that we convert are those that originate on Sifchain.

Not sure what you mean by this:

> Is this a requirement we need in the code or just something the registry indicates?